### ISSUE TITLE
Feature proposal/offer: command line interface

### ISSUE THREAD

**Issue by moreati** [facelessuser/bracex#27]
I've been [seeking a command `x`, like `seq` but doing brace expansion](https://stackoverflow.com/questions/69383322/is-there-command-like-seq-that-performs-full-bash-like-brace-expansion-to-std). It could be used like

```bash
x "{a..z}/{a..z}/{a..z}/{a..z}/{a..z}/{a..z}" | xargs -d $'\n' mkdir -p
```

I've not found a satisfactory `x', so I'd like to offer a PR allowing `python -m bracex` to serve as `x`. Would you be interested?

I imagine the major elements would be
- a new `__main__.py` wrapping `bracex.iexpand()`
- minimal argument parsing using `argparse`
- updates to documentation
- additional tests to maintain code coverage

I totally understand if the answer is no, so please don't feel pressured. I know every new feature creates support costs. As plan B I could release a separate package that wraps bracex, keeping a loose coupling.

Thanks for your consideration, Alex

[1]:

**Comment by moreati** [comment]
I think a command streaming the expansions to stdout would be useful because it wouldn't require O(n**m) memory for n terms, each with m elements. Shells I've tested (bash, zsh, ..) implement brace expansion by trying to allocate the full array at once. For brace expressions with many terms that can exceed available RAM.

**Comment by facelessuser** [comment]
I'm curious, have you tested the differences between `iexpand` and `expand`and verified that it takes less memory?  I only say this as I haven't. Yeah, everything is wrapped up in generators and such, but I've never sat down to verify that generators don't just unravel to be pretty big when spitting out the iterated values. Sure we don't allocate the whole result array, but I've just never sat down to really analyze the results. 

Ultimately, I needed brace expansion, which I pulled off pretty decently,  and the idea to have one as an iterator was nice in theory, but I never proved out whether it gave a substantial advantage one way or the other 🙃.

**Comment by moreati** [comment]
> have you tested the differences between iexpand and expandand verified that it takes less memory?

I've not. I will. Thanks for the heads up.

**Comment by facelessuser** [comment]
While I'm currently uncertain as to whether `iexpand` ends up doing much better as the necessary context may cause a lot of it to unravel, as far as a CLI is concerned, assuming you find the results worth the effort, I'll have to think about it.

@gir-bot remove S: triage
@gir-bot add T: feature, S: needs-decision

**Comment by moreati** [comment]
`bracex.iexpand()` is not O(m**n) with memory. On my laptop
- bash fails at 5 * `{a..z}` (26**5, 11.8 million)
- `bracex.iexpand()` works to at least 6 * `{a..z}` (26**6, 308 million).

```python
# bx
import sys

import bracex

for entry in bracex.iexpand(sys.argv[1], limit=0):
    print(entry)
```
```
$ time bash -c "printf '%s\n' {a..z}/{a..z}/{a..z}/{a..z}/{a..z}"
Killed

real	0m5.676s
user	0m3.673s
sys	0m1.011s
$ time v38/bin/python bx "{a..z}/{a..z}/{a..z}/{a..z}/{a..z}" | wc -l
11881376

real	0m7.901s
user	0m8.003s
sys	0m0.158s
$ time v38/bin/python bx "{a..z}/{a..z}/{a..z}/{a..z}/{a..z}/{a..z}" | wc -l
308915776

real	3m47.018s
user	3m51.033s
sys	0m7.881s
```

**Comment by facelessuser** [comment]
Yep, this actually doesn't surprise me. I suspected, due to context needed when expanding that the generators would probably fully unfold. Add to that the extra overhead that python naturally adds.... You get the idea.

The reason no one iterates is because you can't expand without having all the context. In our case, the attempt to iterate may even add more overhead.

Python is always going to be less efficient than good, optimized C code. This at least confirms what I kind of already suspected. We'd probably be more efficient if we didn't add the overhead of generators, but probably never as good as C.

This has been a useful exercise, but probably didn't give you what you were going for 🙂.

**Comment by facelessuser** [comment]
I should probably, at the very least, set people's expectations for what `iexpand` can actually do in the documents. Like you, I imagine some people came in thinking they'd get a far greater boost, but I know when I finally got iterating working, I was looking at the code very skeptical of how much memory would actually get saved. At the time though, it was good enough to get me what I needed for https://github.com/facelessuser/wcmatch.

**Comment by moreati** [comment]
I think there's a misunderstanding. In my view Bash failed the experiment,
and bracex.iexpand() passed.

I agree python will be slowing it down, vs e.g. Rust, but it's good enough
for me.

**Comment by facelessuser** [comment]
Thanks for confirming that it does what I hoped! Clearly, I was more pessimistic about my own work.

**Comment by facelessuser** [comment]
I think I may be open to a CLI interface. It's probably not something many are specifically looking for, but it should require extremely minimal maintenance.

What do you propose the command be called, and what options?

@gir-bot remove S: needs-decision

**Comment by moreati** [comment]
I was thinking something similar to `seq`
```
usage: python -m bracex [OPTION]... <expression>

Options:
    -s, --seperator=STRING
        use STRING to separate entries (default: \n)
    --help ...
    --version ...
```
I'm not proposing to add anything to $PATH. That could be done later if theirs demand.

**Comment by facelessuser** [comment]
> I'm not proposing to create a new command in $PATH. That could be done later, if there's demand.

This is trivial to do. And I don't think it would conflict with anything. We could do it with very little effort.

Can you elaborate on the `-s` option? Are speaking of how the info is returned? I'm not sure I understand.

**Comment by moreati** [comment]
> This is trivial to do. And I don't think it would conflict with anything. We could do it with very little effort.

I'm aware, but I'd rather not do it from the start because it's easier/cleaner to add something later than to try and remove it because your earlier choice turned out wrong.
 
> Can you elaborate on the `-s` option? Are speaking of how the info is returned? I'm not sure I understand.

E.g.
```
$ python -m bracex "{a..c}/{01..03}"
a/01
a/02
...
c/02
c/03
$ # ^--- implicit newline, like find -print
$ python -m bracex -s ' ' "{a..c}/{01..03}"
a/01 a/02 a/03 b/01 b/02 b/03 c/01 c/02 c/03$ # <--- intentionally no trailing new line, like find -print0
$
$ python -m bracex -s $'\0' "base/{a..c}/{01..03}" | xargs -0 mkdir -p
$ tree base/
base/
├── a
│   ├── 01
...
└── c
    ├── 01
    ├── 02
    └── 03

12 directories, 0 files
```

**Comment by facelessuser** [comment]
Gotcha. Yeah, I'm okay moving forward.

**Comment by moreati** [comment]
Great, thank you for your time. ETA this weekend.

**Comment by moreati** [comment]
There is a memory allocation hiding in `iexpand()`, I just didn't push it far enough. Your suspicions were on the money. I'm investigating.
```
:~/src/bracex [py38] cli(+3/-3)+*± python -mbracex "{a..z}{a..z}{a..z}{a..z}{a..z}{a..z}" | wc -l
308915776
:~/src/bracex [py38] cli(+3/-3)+*± python -mbracex "{a..z}{a..z}{a..z}{a..z}{a..z}{a..z}{a..z}"
Killed
```

**Comment by facelessuser** [comment]
Yeah, it may do better than bash, but I don't think it is completely possible to have it iterate without unfolding context. In some cases, it may do better than others, and in some cases, it may not do that great.

**Comment by moreati** [comment]
Pain is ignorance leaving my head, I hope.
